File: voice-service/app/audio/codec.py
"""G.711 mu-law codec + base64 framing for Plivo Audio Streaming.

Pure Python (Python 3.13+ removed the stdlib `audioop`), no numpy. Plivo streams 8 kHz audio as
base64 frames — mu-law (G.711u) or L16 (16-bit PCM). We convert to/from 16-bit PCM, which is what
IndicConformer STT consumes and Parler-TTS emits.
"""
import base64

_BIAS = 0x84
_CLIP = 32635
# exponent lookup for linear->mu-law (256 entries)
_EXP_LUT = (
    [0, 0, 1, 1] + [2] * 4 + [3] * 8 + [4] * 16 + [5] * 32 + [6] * 64 + [7] * 128
)


def linear_to_ulaw(sample):
    """One signed 16-bit PCM sample -> one mu-law byte (0..255)."""
    sign = 0
    if sample < 0:
        sample = -sample
        sign = 0x80
    if sample > _CLIP:
        sample = _CLIP
    sample += _BIAS
    exponent = _EXP_LUT[(sample >> 7) & 0xFF]
    mantissa = (sample >> (exponent + 3)) & 0x0F
    return (~(sign | (exponent << 4) | mantissa)) & 0xFF


def ulaw_to_linear(u):
    """One mu-law byte -> signed 16-bit PCM sample."""
    u = ~u & 0xFF
    sign = u & 0x80
    exponent = (u >> 4) & 0x07
    mantissa = u & 0x0F
    sample = (((mantissa << 3) + _BIAS) << exponent) - _BIAS
    return -sample if sign else sample


def pcm16_to_ulaw(pcm):
    """bytes of little-endian signed 16-bit PCM -> mu-law bytes."""
    out = bytearray(len(pcm) // 2)
    for i in range(0, len(pcm) - 1, 2):
        s = int.from_bytes(pcm[i:i + 2], "little", signed=True)
        out[i // 2] = linear_to_ulaw(s)
    return bytes(out)


def ulaw_to_pcm16(u):
    """mu-law bytes -> little-endian signed 16-bit PCM bytes."""
    out = bytearray(len(u) * 2)
    for i, b in enumerate(u):
        s = ulaw_to_linear(b)
        out[i * 2:i * 2 + 2] = int(s).to_bytes(2, "little", signed=True)
    return bytes(out)


def decode_frame(payload_b64, audio_format):
    """A Plivo stream frame (base64) -> 16-bit PCM bytes."""
    raw = base64.b64decode(payload_b64)
    if audio_format == "mulaw":
        return ulaw_to_pcm16(raw)
    return raw  # l16 is already PCM16


def encode_frame(pcm, audio_format):
    """16-bit PCM bytes -> base64 payload for a Plivo playback frame."""
    raw = pcm16_to_ulaw(pcm) if audio_format == "mulaw" else pcm
    return base64.b64encode(raw).decode("ascii")

File: voice-worker/src/codec.js
// G.711 mu-law codec + Plivo base64 framing (JS port of voice-service/app/audio/codec.py).
// Everything internal is 8 kHz mono PCM16 (Int16Array).
const BIAS = 0x84, CLIP = 32635;
const EXP_LUT = (() => {
  const a = [0, 0, 1, 1];
  for (const [v, n] of [[2, 4], [3, 8], [4, 16], [5, 32], [6, 64], [7, 128]]) for (let i = 0; i < n; i++) a.push(v);
  return a; // 256 entries
})();

export function linearToUlaw(s) {
  let sign = 0;
  if (s < 0) { s = -s; sign = 0x80; }
  if (s > CLIP) s = CLIP;
  s += BIAS;
  const exponent = EXP_LUT[(s >> 7) & 0xFF];
  const mantissa = (s >> (exponent + 3)) & 0x0F;
  return (~(sign | (exponent << 4) | mantissa)) & 0xFF;
}

export function ulawToLinear(u) {
  u = ~u & 0xFF;
  const sign = u & 0x80;
  const exponent = (u >> 4) & 0x07;
  const mantissa = u & 0x0F;
  const s = (((mantissa << 3) + BIAS) << exponent) - BIAS;
  return sign ? -s : s;
}

export function b64ToBytes(b64) {
  const bin = atob(b64);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export function bytesToB64(bytes) {
  let bin = "";
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]);
  return btoa(bin);
}

// A Plivo inbound media frame (base64) -> Int16Array PCM16 @ 8k.
export function decodeFrame(b64, fmt) {
  const raw = b64ToBytes(b64);
  if (fmt === "mulaw") {
    const out = new Int16Array(raw.length);
    for (let i = 0; i < raw.length; i++) out[i] = ulawToLinear(raw[i]);
    return out;
  }
  // l16: little-endian PCM16 bytes
  const out = new Int16Array(raw.length >> 1);
  for (let i = 0; i < out.length; i++) out[i] = (raw[i * 2] | (raw[i * 2 + 1] << 8)) << 16 >> 16;
  return out;
}

// Int16Array PCM16 @ 8k -> base64 payload for a Plivo playAudio frame.
export function encodeFrame(pcm, fmt) {
  if (fmt === "mulaw") {
    const u = new Uint8Array(pcm.length);
    for (let i = 0; i < pcm.length; i++) u[i] = linearToUlaw(pcm[i]);
    return bytesToB64(u);
  }
  const b = new Uint8Array(pcm.length * 2);
  for (let i = 0; i < pcm.length; i++) { b[i * 2] = pcm[i] & 0xFF; b[i * 2 + 1] = (pcm[i] >> 8) & 0xFF; }
  return bytesToB64(b);
}

// Mean |amplitude| of a PCM16 frame — cheap energy for silence detection.
export function rms(pcm) {
  if (!pcm.length) return 0;
  let total = 0;
  for (let i = 0; i < pcm.length; i++) total += Math.abs(pcm[i]);
  return (total / pcm.length) | 0;
}

// Parse a WAV (from Sarvam TTS) -> { pcm: Int16Array, rate }. Minimal RIFF reader.
export function wavToPcm(bytes) {
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  let rate = 8000, off = 12; // skip RIFF header
  let dataOff = -1, dataLen = 0;
  while (off + 8 <= bytes.length) {
    const id = String.fromCharCode(bytes[off], bytes[off + 1], bytes[off + 2], bytes[off + 3]);
    const size = dv.getUint32(off + 4, true);
    if (id === "fmt ") rate = dv.getUint32(off + 12, true);
    else if (id === "data") { dataOff = off + 8; dataLen = size; break; }
    off += 8 + size + (size & 1);
  }
  if (dataOff < 0) return { pcm: new Int16Array(0), rate };
  const n = dataLen >> 1;
  const pcm = new Int16Array(n);
  for (let i = 0; i < n; i++) pcm[i] = dv.getInt16(dataOff + i * 2, true);
  return { pcm, rate };
}

// Linear resample (anti-aliased box filter on downsample) — mirrors app/audio/resample.py.
export function resample(pcm, srcHz, dstHz) {
  if (srcHz === dstHz || !pcm.length) return pcm;
  const nOut = Math.floor(pcm.length * dstHz / srcHz);
  const out = new Int16Array(nOut);
  if (dstHz < srcHz) {
    const win = srcHz / dstHz;
    for (let i = 0; i < nOut; i++) {
      let a = (i * win) | 0, b = ((i + 1) * win) | 0;
      if (b <= a) b = a + 1;
      if (b > pcm.length) b = pcm.length;
      let sum = 0;
      for (let j = a; j < b; j++) sum += pcm[j];
      out[i] = (sum / (b - a)) | 0;
    }
  } else {
    const ratio = dstHz / srcHz;
    for (let i = 0; i < nOut; i++) {
      const pos = i / ratio, j = pos | 0, frac = pos - j;
      out[i] = (j + 1 < pcm.length) ? (pcm[j] * (1 - frac) + pcm[j + 1] * frac) | 0 : pcm[pcm.length - 1];
    }
  }
  return out;
}
